### PR TITLE
feat: custom token renderer

### PR DESCRIPTION
**Pull request by 0xble**
## Motivation

Party Cards are the only way to visually represent a Party. But the only relationship between a Party and its Party Card is the name. Everything else is PartyDAO branded. Many DAOs and communities in the space have unique art and would not use Party because they cannot express themselves. There are a number of Parties that would be started if only they could use a custom image to replace the Party Card.

## Solution

Support displaying custom metadata (title, image, description, etc.) for a Party’s NFT memberships through a new `MetadataRegistry` contract. This data is set by the Party creator and can be voted on to be updated by Party members with the addition of a new `CustomizeMetadataProposal`.

### CODE REVIEW

File: contracts/renderers/MetadataRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    function setProvider(IMetadataProvider provider) public {
+        getProvider[msg.sender] = provider;
+    }
+
+    /// @notice Get the metadata for a Party instance.
+    /// @param instance The address of the instance.
+    /// @return metadata The encoded metadata.
+    function getMetadata(address instance) external view returns (bytes memory metadata) {
+        IMetadataProvider provider = getProvider[instance];
+        return address(provider) != address(0) ? provider.getMetadata(instance) : metadata;
```
[the remark this answers is not in the document]
Comment (0xble) [reply contracts/renderers/MetadataRegistry.sol:32]: Yeah idk why I did that lol

File: contracts/party/PartyFactory.sol [hunk 2]
```diff
@@ -6,11 +6,22 @@ import "../utils/Proxy.sol";
 
 import "./Party.sol";
 import "./IPartyFactory.sol";
+import "../renderers/MetadataRegistry.sol";
+import "../renderers/MetadataProvider.sol";
```
Comment (arr00) [inline contracts/party/PartyFactory.sol:10]: named imports

File: contracts/proposals/CustomizeMetadataProposal.sol [hunk 3]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import "./IProposalExecutionEngine.sol";
+import "../globals/LibGlobals.sol";
+import "../renderers/MetadataRegistry.sol";
+import "../renderers/MetadataProvider.sol";
```
Comment (arr00) [inline contracts/proposals/CustomizeMetadataProposal.sol:7]: named imports

File: contracts/proposals/ProposalExecutionEngine.sol [hunk 4]
```diff
@@ -15,6 +15,7 @@ import "./ProposalStorage.sol";
 import "./DistributeProposal.sol";
 import "./AddAuthorityProposal.sol";
 import "./OperatorProposal.sol";
+import "./CustomizeMetadataProposal.sol";
```
Comment (arr00) [inline contracts/proposals/ProposalExecutionEngine.sol:18]: named imports

File: contracts/renderers/MetadataRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    /// @notice Whether or not an address is a registar that can set the
+    ///         provider and metadata for another instance. If registrar is set
+    ///         true for `address(1)`, the address is a universal registar and
+    ///         can set data for any instance.
+    /// @dev Registrars' ability to set metadata for another instance must also be
+    ///      supported by the metadata provider used by that instance, indicated by
+    ///      `IMetadataProvider.supportsRegistrars()`.
+    mapping(address registrar => mapping(address instance => bool canSetData)) private _isRegistrar;
+
+    constructor(IGlobals globals, address[] memory registrars) {
```
Comment (arr00) [inline contracts/renderers/MetadataRegistry.sol:32]: natspec?

File: contracts/renderers/PartyNFTRenderer.sol [hunk 6]
```diff
@@ -6,6 +6,7 @@ import "../utils/vendor/Strings.sol";
 import "../utils/vendor/Base64.sol";
 
 import "./RendererBase.sol";
+import "./MetadataRegistry.sol";
```
Comment (arr00) [inline contracts/renderers/PartyNFTRenderer.sol:9]: named import

**Review by arr00 — COMMENTED** [review]
Some comments so far. Also need at add `_executeCustomizeMetadata` and integration test for `CustomizeMetadataProposal`

File: test/renderers/MetadataRegistry.t.sol [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+        for (uint256 i; i < instances.length; ++i) {
+            vm.prank(registrar);
+            vm.expectEmit(true, true, true, true);
+            emit ProviderSet(instances[i], provider);
+            registry.setProvider(instances[i], provider);
+
+            assertTrue(registry.getProvider(instances[i]) == provider);
+        }
+    }
```
Comment (arr00) [inline test/renderers/MetadataRegistry.t.sol:157]: could we add a test `test_setProvider_asInstance`?

**Review by arr00 — APPROVED** [review]
Approving from a security stand point. Graphic rendering review was brief.

**Comment by 0xble** [comment]
Didn't notice when I merged this some new tests were failing, I'll fix that

**Comment by arr00** [comment]
Looks good

**Pull request merged**